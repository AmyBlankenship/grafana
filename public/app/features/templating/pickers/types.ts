import { VariableModel } from '../variable';

export interface VariablePickerProps<Model extends VariableModel = VariableModel> {
  variable: Model;
}
