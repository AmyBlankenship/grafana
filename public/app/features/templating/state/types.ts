import { VariableModel } from '../variable';
import { VariablesState } from './variablesReducer';

export const EMPTY_UUID = '00000000-0000-0000-0000-000000000000';
export const ALL_VARIABLE_TEXT = 'All';
export const ALL_VARIABLE_VALUE = '$__all';
export const NONE_VARIABLE_TEXT = 'None';
export const NONE_VARIABLE_VALUE = '';

// TODO: move to pickers/types?
export interface VariablePickerProps<Model extends VariableModel = VariableModel> {
  variable: Model;
}

export const getInstanceState = <Model extends VariableModel = VariableModel>(state: VariablesState, uuid: string) => {
  return state[uuid] as Model;
};
