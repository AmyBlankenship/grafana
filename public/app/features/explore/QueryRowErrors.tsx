import React, { FunctionComponent } from 'react';
import { ErrorContainer } from './ErrorContainer';
import { DataQueryError } from '@grafana/data';

export interface Props {
  queryErrors: DataQueryError[];
}

export const QueryRowErrors: FunctionComponent<Props> = (props: Props) => {
  const { queryErrors } = props;

  const errors = queryErrors.map(error => {
    return {
      ...error,
      refId: null,
    };
  });

  return <ErrorContainer queryErrors={errors} />;
};
