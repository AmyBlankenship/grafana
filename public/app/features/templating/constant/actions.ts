import {
  toVariablePayload,
  validateVariableSelectionState,
  VariableIdentifier,
  VariablePayload,
} from '../state/actions';
import { ThunkResult } from 'app/types';
import { createAction } from '@reduxjs/toolkit';

export const createConstantOptionsFromQuery = createAction<VariablePayload>('templating/createCustomOptionsFromQuery');

export const updateConstantVariableOptions = (identifier: VariableIdentifier): ThunkResult<void> => {
  return async dispatch => {
    await dispatch(createConstantOptionsFromQuery(toVariablePayload(identifier)));
    await dispatch(validateVariableSelectionState(identifier));
  };
};
