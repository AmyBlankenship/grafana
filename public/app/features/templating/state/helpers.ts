import { EMPTY_UUID } from './types';
import { QueryVariableModel, VariableHide, VariableModel } from '../variable';
import { initialQueryVariableModelState } from '../query/reducer';
import { VariablesState } from './variablesReducer';

export const getVariableState = (
  noOfVariables: number,
  inEditorIndex = -1,
  includeEmpty = false
): Record<string, VariableModel> => {
  const variables: Record<string, VariableModel> = {};

  for (let index = 0; index < noOfVariables; index++) {
    variables[index] = {
      uuid: index.toString(),
      type: 'query',
      name: `Name-${index}`,
      hide: VariableHide.dontHide,
      index,
      label: `Label-${index}`,
      skipUrlSync: false,
    };
  }

  if (includeEmpty) {
    variables[EMPTY_UUID] = {
      uuid: EMPTY_UUID,
      type: 'query',
      name: `Name-${EMPTY_UUID}`,
      hide: VariableHide.dontHide,
      index: noOfVariables,
      label: `Label-${EMPTY_UUID}`,
      skipUrlSync: false,
    };
  }

  return variables;
};

export const getVariableTestContext = (variableOverrides: Partial<QueryVariableModel> = {}) => {
  const defaultVariable = {
    ...initialQueryVariableModelState,
    uuid: '0',
    index: 0,
    name: '0',
  };
  const variable = { ...defaultVariable, ...variableOverrides };
  const initialState: VariablesState = {
    '0': variable,
  };

  return { initialState };
};
