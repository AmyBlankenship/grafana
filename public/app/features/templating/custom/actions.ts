import { toVariablePayload, validateVariableSelectionState, VariableIdentifier } from '../state/actions';
import { ThunkResult } from 'app/types';
import { createCustomOptionsFromQuery } from './reducer';

export const updateCustomVariableOptions = (identifier: VariableIdentifier): ThunkResult<void> => {
  return async dispatch => {
    await dispatch(createCustomOptionsFromQuery(toVariablePayload(identifier)));
    await dispatch(validateVariableSelectionState(identifier));
  };
};
