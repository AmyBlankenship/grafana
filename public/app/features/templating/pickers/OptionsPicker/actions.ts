import debounce from 'lodash/debounce';
import { StoreState, ThunkDispatch, ThunkResult } from 'app/types';
import { containsSearchFilter, VariableOption, VariableWithMultiSupport, VariableWithOptions } from '../../variable';
import { toVariablePayload, setCurrentVariableValue } from '../../state/actions';
import { variableAdapters } from '../../adapters';
import { getVariable } from '../../state/selectors';
import { NavigationKey } from '../shared/types';
import {
  hideOptions,
  OptionsPickerState,
  updateOptionsFromSearch,
  updateSearchQuery,
  updateOptionsAndFilter,
  toggleOption,
  moveOptionsHighlight,
} from './reducer';

export const navigateOptions = (key: NavigationKey, clearOthers: boolean): ThunkResult<void> => {
  return (dispatch, getState) => {
    if (key === NavigationKey.cancel) {
      return dispatch(commitChangesToVariable());
    }

    if (key === NavigationKey.select) {
      return dispatch(toggleOptionByHighlight(clearOthers));
    }

    if (key === NavigationKey.selectAndClose) {
      dispatch(toggleOptionByHighlight(clearOthers));
      return dispatch(commitChangesToVariable());
    }

    if (key === NavigationKey.moveDown) {
      return dispatch(moveOptionsHighlight(1));
    }

    if (key === NavigationKey.moveUp) {
      return dispatch(moveOptionsHighlight(-1));
    }

    return undefined;
  };
};

export const filterOrSearchOptions = (searchQuery: string): ThunkResult<void> => {
  return async (dispatch, getState) => {
    const { uuid } = getState().templating.optionsPicker;
    const { query, options } = getVariable<VariableWithOptions>(uuid, getState());
    dispatch(updateSearchQuery(searchQuery));

    if (containsSearchFilter(query)) {
      return searchForOptionsWithDebounce(dispatch, getState, searchQuery);
    }
    return dispatch(updateOptionsAndFilter(options));
  };
};

export const commitChangesToVariable = (): ThunkResult<void> => {
  return async (dispatch, getState) => {
    const picker = getState().templating.optionsPicker;
    const existing = getVariable<VariableWithMultiSupport>(picker.uuid, getState());
    const current = mapToCurrent(picker);

    dispatch(setCurrentVariableValue(toVariablePayload(existing, current)));
    const updated = getVariable<VariableWithMultiSupport>(picker.uuid, getState());

    if (existing.current.text === updated.current.text) {
      return dispatch(hideOptions());
    }

    const adapter = variableAdapters.get(updated.type);
    await adapter.setValue(updated, updated.current, true);
    return dispatch(hideOptions());
  };
};

export const toggleOptionByHighlight = (clearOthers: boolean): ThunkResult<void> => {
  return (dispatch, getState) => {
    const { uuid, highlightIndex } = getState().templating.optionsPicker;
    const variable = getVariable<VariableWithMultiSupport>(uuid, getState());
    const option = variable.options[highlightIndex];
    dispatch(toggleOption({ option, forceSelect: false, clearOthers }));
  };
};

const searchForOptions = async (dispatch: ThunkDispatch, getState: () => StoreState, searchQuery: string) => {
  try {
    const { uuid } = getState().templating.optionsPicker;
    const existing = getVariable<VariableWithOptions>(uuid, getState());

    const adapter = variableAdapters.get(existing.type);
    await adapter.updateOptions(existing, searchQuery);

    const updated = getVariable<VariableWithOptions>(uuid, getState());
    dispatch(updateOptionsFromSearch(updated.options));
  } catch (error) {
    console.error(error);
  }
};

const searchForOptionsWithDebounce = debounce(searchForOptions, 500);

function mapToCurrent(picker: OptionsPickerState): VariableOption {
  const { options, searchQuery, multi } = picker;

  if (options.length === 0 && searchQuery && searchQuery.length > 0) {
    return { text: searchQuery, value: searchQuery, selected: false };
  }

  if (!multi) {
    return options.find(o => o.selected);
  }

  const texts: string[] = [];
  const values: string[] = [];

  for (const option of options) {
    if (!option.selected) {
      continue;
    }

    texts.push(option.text.toString());
    values.push(option.value.toString());
  }

  return {
    value: values,
    text: texts.join(' + '),
    tags: picker.tags.filter(t => t.selected),
    selected: true,
  };
}
