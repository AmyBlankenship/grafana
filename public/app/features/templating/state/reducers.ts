import { combineReducers } from '@reduxjs/toolkit';
import { optionsPickerReducer, OptionsPickerState } from '../pickers/OptionsPicker/reducer';
import { variableEditorReducer, VariableEditorState } from '../editor/reducer';
import { uuidInEditorReducer } from './uuidInEditorReducer';
import { variablesReducer } from './variablesReducer';
import { VariableModel } from '../variable';

export interface TemplatingState {
  variables: Record<string, VariableModel>;
  optionsPicker: OptionsPickerState;
  editor: VariableEditorState;
  uuidInEditor: string | null;
}

export default {
  templating: combineReducers({
    uuidInEditor: uuidInEditorReducer,
    optionsPicker: optionsPickerReducer,
    editor: variableEditorReducer,
    variables: variablesReducer,
  }),
};
