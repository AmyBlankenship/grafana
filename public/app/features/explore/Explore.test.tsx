import React from 'react';
import { DataSourceApi, LoadingState, ExploreMode, toUtc, DataQueryError, DataQueryRequest } from '@grafana/data';
import { ExploreId } from 'app/types/explore';
import { shallow } from 'enzyme';
import { Explore, ExploreProps } from './Explore';
import { scanStopAction } from './state/actionTypes';
import { toggleGraph } from './state/actions';

const setup = (renderMethod: any, propOverrides?: object) => {
  const props: ExploreProps = {
    changeSize: jest.fn(),
    datasourceInstance: {
      meta: {
        metrics: true,
        logs: true,
      },
      components: {
        ExploreStartPage: {},
      },
    } as DataSourceApi,
    datasourceMissing: false,
    exploreId: ExploreId.left,
    initializeExplore: jest.fn(),
    initialized: true,
    modifyQueries: jest.fn(),
    update: {
      datasource: false,
      queries: false,
      range: false,
      mode: false,
      ui: false,
    },
    refreshExplore: jest.fn(),
    scanning: false,
    scanRange: {
      from: '0',
      to: '0',
    },
    scanStart: jest.fn(),
    scanStopAction: scanStopAction,
    setQueries: jest.fn(),
    split: false,
    queryKeys: [],
    initialDatasource: 'test',
    initialQueries: [],
    initialRange: {
      from: toUtc('2019-01-01 10:00:00'),
      to: toUtc('2019-01-01 16:00:00'),
      raw: {
        from: 'now-6h',
        to: 'now',
      },
    },
    mode: ExploreMode.Metrics,
    initialUI: {
      showingTable: false,
      showingGraph: false,
      showingLogs: false,
    },
    isLive: false,
    syncedTimes: false,
    updateTimeRange: jest.fn(),
    graphResult: [],
    loading: false,
    absoluteRange: {
      from: 0,
      to: 0,
    },
    showingGraph: false,
    showingTable: false,
    timeZone: 'UTC',
    onHiddenSeriesChanged: jest.fn(),
    toggleGraph: toggleGraph,
    queryResponse: {
      state: LoadingState.Done,
      series: [],
      request: {} as DataQueryRequest,
      error: {} as DataQueryError,
      timeRange: {
        from: toUtc('2019-01-01 10:00:00'),
        to: toUtc('2019-01-01 16:00:00'),
        raw: {
          from: 'now-6h',
          to: 'now',
        },
      },
    },
    originPanelId: 1,
    addQueryRow: jest.fn(),
  };

  Object.assign(props, propOverrides);
  return renderMethod(<Explore {...props} />);
};

describe('Explore', () => {
  it('should render component', () => {
    const wrapper = setup(shallow);
    expect(wrapper).toMatchSnapshot();
  });
});
