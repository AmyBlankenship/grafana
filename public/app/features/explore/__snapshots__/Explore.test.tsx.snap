// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explore should render component 1`] = `
<div
  className="explore"
>
  <Connect(UnConnectedExploreToolbar)
    exploreId="left"
    onChangeTime={[Function]}
  />
  <div
    className="explore-container"
  >
    <QueryRows
      exploreEvents={
        Emitter {
          "emitter": EventEmitter {
            "_events": Object {},
            "_eventsCount": 0,
          },
        }
      }
      exploreId="left"
      queryKeys={Array []}
    />
    <div
      className="gf-form"
    >
      <button
        aria-label="Add row button"
        className="gf-form-label gf-form-label--btn css-o41ti2"
        disabled={false}
        onClick={[Function]}
      >
        <i
          className="fa fa-fw fa-plus icon-margin-right"
        />
        <span
          className="btn-title"
        >
          Â Add query
        </span>
      </button>
    </div>
    <Component
      queryError={Object {}}
    />
    <AutoSizer
      disableHeight={true}
      disableWidth={false}
      onResize={[Function]}
      style={Object {}}
    >
      <Component />
    </AutoSizer>
  </div>
</div>
`;
